cmd/registry/controller/controller.go

```diff
@@ -232,73 +233,51 @@ func generateCreateActions(
 	generatedResource *rpc.GeneratedResource,
 	visited map[string]bool) ([]*Action, error) {
 
+	var parentList []resourceInstance
+
 	parsedResourcePattern, err := parseResourcePattern(resourcePattern)
 	if err != nil {
 		return nil, err
 	}
-	parentName := parsedResourcePattern.Parent()
-
-	// If parent is a project, we can't list projects since this is registry client command.
-	// Since the manifest definition is scoped  only for a particular project,
-	// there will be only one target resource in this case.
-	// There are two cases where this might happen:
-	// 1. Target resource is a project level artifact "projects/demo/locations/global/artifacts/serach-index"
-	//    extracted parent will be "projects/demo/locations/global"
-	// 1. Target resource is an api "projects/demo/locations/global/apis/petstore"
-	//    extracted parent will be "projects/demo/locations/global"
-	if strings.HasSuffix(parentName, "locations/global") {
+
+	parentName := parsedResourcePattern.ParentName()
+	switch parentName.(type) {
+	case projectName:
+		// If parent is a project, we can't list projects since this is registry client command.
+		// Since the manifest definition is scoped  only for a particular project,
+		// there will be only one target resource in this case.
+		// There are two cases where this might happen:
+		// 1. Target resource is a project level artifact "projects/demo/locations/global/artifacts/serach-index"
+		//    extracted parent will be "projects/demo/locations/global"
+		// 1. Target resource is an api "projects/demo/locations/global/apis/petstore"
+		//    extracted parent will be "projects/demo/locations/global"
 		// Return if this parent was already visited.
-		if visited[parentName] {
+		if visited[parentName.String()] {
 			return nil, nil
 		}
-
-		// Since the GeneratedResource is non-existent here,
-		// we will have to derive the exact name of the target resource.
-		targetResourceName, err := resourceNameFromParent(resourcePattern, parentName)
-		if err != nil {
-			return nil, fmt.Errorf("Cannot generate target resourceName to be created. Error: %s", err)
+		parentList = []resourceInstance{
+			projectResource{
+				projectName: parentName,
+			},
 		}
 
-		takeAction, err := needsCreate(
-			targetResourceName,
-			dependencyMaps,
-			generatedResource,
-		)
-
+	default:
+		// If parent resource is not a project, then go through all the non-visited parents.
+		visited_json, err := json.Marshal(visited)
 		if err != nil {
-			return nil, err
-		} else if !takeAction {
-			return nil, nil
+			return nil, fmt.Errorf("Internal error: Invalid visited map, %s", err)
 		}
-
-		cmd, err := generateCommand(generatedResource.Action, targetResourceName.String())
+		// Construct a filter to fetch only non-visited parents.
+		filter := fmt.Sprintf("%s.all(n, !(name.contains(n)))", visited_json)
```

Comment:
What's happening here? I think this is worth explaining in more detail in the code comment.

Reply:
Added more explanation, let me know if it is clear enough to understand.